packages/instrumentations/src/index.ts

```diff
@@ -18,136 +18,212 @@ import {
+  "@opentelemetry/instrumentation-mysql2": "MySQL2Instrumentation",
+  "@opentelemetry/instrumentation-mysql": "MySQLInstrumentation",
+  "@opentelemetry/instrumentation-nestjs-core": "NestInstrumentation",
+  "@opentelemetry/instrumentation-net": "NetInstrumentation",
+  "@opentelemetry/instrumentation-pg": "PgInstrumentation",
+  "@opentelemetry/instrumentation-pino": "PinoInstrumentation",
+  "@opentelemetry/instrumentation-redis": "RedisInstrumentation",
+  "@opentelemetry/instrumentation-redis-4": "RedisInstrumentation",
+  "@opentelemetry/instrumentation-restify": "RestifyInstrumentation",
+  "@opentelemetry/instrumentation-router": "RouterInstrumentation",
+  "@opentelemetry/instrumentation-socket.io": "SocketIoInstrumentation",
+  "@opentelemetry/instrumentation-tedious": "TediousInstrumentation",
+  "@opentelemetry/instrumentation-winston": "WinstonInstrumentation",
+}
+
+// map of resource detector package names to the names of their exported detectors
+const RESOURCE_DETECTOR_NAMES: Record<string, string[]> = {
+  "@opentelemetry/resource-detector-aws": ["awsEc2Detector"],
+  "@opentelemetry/resource-detector-azure": ["azureAppServiceDetector"],
+  "@opentelemetry/resource-detector-container": ["containerDetector"],
+  "@opentelemetry/resources": [
+    "envDetectorSync",
+    "hostDetectorSync",
+    "osDetectorSync",
+    "processDetectorSync",
+  ],
 }
+
 export type InstrumentationConfigMap = {
-  [I in keyof typeof INSTRUMENTATIONS]?: (typeof INSTRUMENTATIONS)[I] extends new (
-    config: infer C,
-    ...args: unknown[]
-  ) => unknown
+  [I in keyof InstrumentationTypes]?: InstrumentationTypes[I] extends {
+    setConfig(config: infer C): unknown
+  }
     ? C
     : never
 }
 
 export function getInstrumentations(
   configs: InstrumentationConfigMap,
   defaultDisabled: boolean,
-): Instrumentation[] {
-  const instrumentations: Instrumentation[] = []
-  const c: Record<string, InstrumentationConfig> = configs
+): Instrumentation[] | Promise<Instrumentation[]> {
+  const allConfigs = Object.assign(
+    // one entry per available instrumentations
+    Object.fromEntries(
+      Object.keys(INSTRUMENTATION_NAMES).map((name) => [name, {}]),
+    ),
+    // override defaults using the entries present in the config while
+    // filtering out entries from the config that are not available
+    Object.fromEntries(
+      Object.entries(configs).filter(([name]) =>
+        Object.keys(INSTRUMENTATION_NAMES).includes(name),
+      ),
+    ),
+  )
 
-  for (const [name, Class] of Object.entries<
-    new (config?: InstrumentationConfig) => Instrumentation
-  >(INSTRUMENTATIONS)) {
-    let config = c[name]
+  const instrumentations = Object.entries(allConfigs)
+    .filter(([, config]: [unknown, InstrumentationConfig]) => {
+      // explicitly set "enabled" to false if that's the default
+      if (defaultDisabled) config.enabled ??= false
+      // filter out disabled instrumentations
+      return config.enabled !== false
+    })
```

Comment:
It seems we currently set `instrumentationsDefaultDisabled: boolean.default(IS_SERVERLESS)`, does this mean no instrumentation will be enabled in lambda?

I'm confused how the upstream https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-aws-lambda instrumentation (and the aws-sdk instrumentation) are available in our test lambda functions if no instrumentation is enabled (maybe the existing layers had them all enabled?).  Also, the thing I mentioned about the upstream lambda instrumentation ignoring [.mjs handler filename](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/node/opentelemetry-instrumentation-aws-lambda/src/instrumentation.ts#L116-L128) -- not seeing how that's addressed in our custom distro, can you point me to how this works?

Reply:
The AWS instrumentation configs are patched to take precedence over the defaults https://github.com/solarwinds/apm-js/blob/main/packages/sdk/src/patches/aws-lambda.ts

I'm actually not sure how the `.mjs` loading works given that code, I just know that it does work. Will look into what's going on there.